# Neural Network Type Classification | TMNIST

Name: Anurag Gher

NUID: 002845719

# Introduction :
The Typography MNIST (TMNIST) dataset contains over 281,000 MNIST-style images of 94 alphabetic characters, including digits, lowercase and uppercase letters, and symbols. Each character is associated with a unique font style. The dataset is provided in a CSV format with font names, character labels, and pixel values. TMNIST showcases typography diversity and is valuable for exploring character recognition, font design, and cognitive science. It is utilized in projects like Warhol.ai Computational Creativity and Cognitive Type, offering insights into typographic aesthetics and visual cognition.

# About the dataset:
The Typography MNIST (TMNIST) dataset comprises 281,000+ images resembling handwritten characters, including uppercase and lowercase letters, digits, and special symbols like '@', '!', and '#'.


{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\', ']', '^', '_', '`', '{', '|', '}', '~'}


 Organized in a CSV file, each row contains image details like font style and character label. Images are represented as 28x28 pixel grids, with grayscale intensities encoded. This dataset is vital for tasks like character recognition and font analysis.


Key Points:
*   Diverse Characters: 94 characters included, covering alphanumeric and special symbol.
*   File Format: Stored in a CSV file for easy access.
*   Dataset Size: Over 281,000 images provide substantial training data.

Overall, the TMNIST dataset is essential for character recognition and font analysis tasks.



Kaggle Link: https://www.kaggle.com/nikbearbrown/tmnist-alphabet-94-characters/

# Imports

**First we will import all the necessary libraries**

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models
from sklearn.model_selection import train_test_split
import numpy as np
import matplotlib.pyplot as plt


**TensorFlow:** It's for building and training machine learning models, especially deep neural networks.

**scikit-learn (sklearn):** It provides tools for data preprocessing, modeling, and evaluation, making it easy to work with machine learning algorithms.

**Matplotlib:** This library helps you visualize data and model performance through various plots and charts.

In [ ]:
import pandas as pd
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

Importing necessary layers and modules from Keras for building a **convolutional neural network (CNN)**. With these libraries, you have everything needed to preprocess data, create neural network models, and incorporate diverse layers like convolutional, pooling, flattening, dense, and dropout layers. These elements are commonly utilized in deep learning applications such as image classification.

# Loading and Transforming the dataset

In [ ]:
#Load data
data = pd.read_csv('TMNIST_Data.csv')

I've imported the Typography MNIST (TMNIST) dataset, encompassing more than 281,000 images portraying 94 different alphabetic characters. These images come in diverse font styles and are represented as 28x28 grayscale pixel arrays.

In [ ]:
data.head()

names  labels  1  2  3  4  5  6  7  8  ...  775  776  \
0          GrandHotel-Regular       2  0  0  0  0  0  0  0  0  ...    0    0   
1  EncodeSansCondensed-Medium       8  0  0  0  0  0  0  0  0  ...    0    0   
2              Varela-Regular       4  0  0  0  0  0  0  0  0  ...    0    0   
3              ArefRuqaa-Bold       3  0  0  0  0  0  0  0  0  ...    0    0   
4           KulimPark-Regular       1  0  0  0  0  0  0  0  0  ...    0    0   

   777  778  779  780  781  782  783  784  
0    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    0  
2    0    0    0    0    0    0    0    0  
3    0    0    0    0    0    0    0    0  
4    0    0    0    0    0    0    0    0  

[5 rows x 786 columns]

The **`head()`** method in pandas provides a brief overview of the initial rows of a DataFrame, typically displaying the first five rows by default. This concise summary offers a glimpse into the dataset's structure and content, aiding in quick data inspection and exploration without having to review the entire dataset.

In [ ]:
#Dataset Characteristics
print(f"The Shape of the Dataset is: {data.shape}")
print(f"Number of Samples: {data.shape[0]}")
print(f"Number of Features: {data.shape[1]}")

The Shape of the Dataset is: (29900, 786)
Number of Samples: 29900
Number of Features: 786


The output provides comprehensive details about the dataset:

- "The Shape of the Dataset is: (29900, 786)": This indicates that the dataset comprises 29,900 samples (rows) and 786 features (columns), showcasing its size and dimensionality.
- "Number of Samples: 29900": This confirms the total number of samples present in the dataset, reflecting the quantity of individual data points available for analysis.
- "Number of Features: 786": This indicates the total number of features or attributes present in the dataset, representing the various characteristics or variables associated with each sample.

In [ ]:
# Number of unique character in the Dataframe
num_unique_chars = len(data.labels.unique())
print(f"Number of unique character present in the Dataset: {num_unique_chars}")

Number of unique character present in the Dataset: 10


The provided code snippet calculates the number of unique characters present in the DataFrame:

"**Number of unique character present in the Dataset: 10**": This output signifies that there are 10 unique characters within the dataset, offering insight into the diversity or variety of characters represented in the data.

In [ ]:
data['labels'].unique()

array([2, 8, 4, 3, 1, 6, 5, 9, 7, 0])

The code snippet extracts and displays the unique values present in the 'labels' column of the DataFrame:

- The output **"[2, 8, 4, 3, 1, 6, 5, 9, 7, 0]"** indicates an array containing the unique values found in the 'labels' column.
- Each unique value represents a distinct category or class within the dataset. In this case, the 'labels' column likely represents categorical data with 10 unique categories ranging from 0 to 9.

In [ ]:
# Normalize pixel values to the range [0, 1]
data.iloc[:, 2:] /= 255.0

In the given code, pixel values of the dataset's images are being normalized to a range of 0 to 1. This is achieved by dividing the pixel values, which originally range from 0 to 255, by 255.0. Normalizing pixel values is crucial for machine learning models, particularly neural networks, as it ensures uniform and standardized input data. This normalization process facilitates better convergence during model training and enhances overall performance.

In [ ]:
# Reshape the image data for convolutional neural network (CNN) input
X = data.iloc[:, 2:].values.reshape(-1, 28, 28, 1)

The provided code snippet is responsible for transforming the image data into a format compatible with Convolutional Neural Networks (CNNs). The resulting shape, (-1, 28, 28, 1), indicates that each image is now represented as a 28x28 pixel grid with a single color channel, indicating grayscale. This reshaping aligns with the input specifications of CNNs, commonly used for image classification tasks. Introducing the additional dimension accommodates the single-channel characteristic of grayscale images, preparing the data for further model training.

In [ ]:
# One-hot encode the labels
y = pd.get_dummies(data['labels']).values

The given code is responsible for converting categorical labels in the dataset into a one-hot encoded format. This transformation represents each unique label as a binary vector, where only one element is set to 1, indicating the presence of a particular class. Such encoding is vital for neural network training, especially in classification scenarios, as it prepares categorical labels for effective model training and evaluation.

# Splitting the data into training and testing sets

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


The code segment splits a dataset into training and testing subsets using scikit-learn's `train_test_split` function. It separates input features and corresponding labels, allocating 20% of the data for testing and the remaining 80% for training. The `random_state=42` parameter ensures reproducibility by fixing the random seed for the split.We are using 20% of the data for testing, while the remaining 80% will be used for training the neural network.

In [ ]:
num_typefaces = len(data['labels'].unique())
print(num_typefaces)

10


The provided code determines the total number of unique typefaces within the dataset. This information is pivotal for setting up the output layer of the neural network, ensuring it aligns with the dataset's unique typefaces. As evident from the output, the dataset comprises 10 distinct typefaces.

# Defining the neural network architecture

In [ ]:
model = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(num_typefaces, activation='softmax')
])

This code outlines the structure of a Convolutional Neural Network (CNN) using TensorFlow and Keras. The model is defined as a sequential stack of layers, comprising convolutional layers with Rectified Linear Unit (ReLU) activation, followed by max-pooling layers for spatial dimension reduction. A flattened layer is included to convert 2D feature maps into a 1D vector. Additionally, two fully connected (dense) layers are incorporated, with ReLU activation in the first dense layer and softmax activation in the output layer. The input shape is configured as (28, 28, 1) to match the dimensions of input images. The number of units in the output layer is determined by the variable num_typefaces, ensuring the network produces probabilities for each unique typeface label. This architecture is well-suited for tasks involving image classification.

In [ ]:
# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

- This code segment compiles the model, configuring it for training, by employing the Adam optimizer, categorical crossentropy loss function, and accuracy metric.
- Compiling the model precedes training and is crucial for its setup.
- The Adam optimizer, known for its adaptive learning rates, is favored for optimization.
- Categorical crossentropy serves as the loss function, fitting for multi-class classification problems with softmax activation.
- The objective during training is to minimize categorical crossentropy loss, while accuracy is employed as a metric to gauge model performance on training and validation datasets.

# Model Summary

In [ ]:
# Display the model summary
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 13, 13, 32)        0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 flatten_1 (Flatten)         (None, 1600)              0         
                                                                 
 dense_2 (Dense)             (None, 128)              

- The model summary presents a thorough outline of the neural network architecture, delineating each layer's type, output shape, and parameter count.
- This summary serves as a valuable resource for grasping the network's structure and parameterization, aiding in efficient debugging and optimization of the model.

In [ ]:
# Train the model
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))

Epoch 1/10
748/748 [==============================] - 23s 29ms/step - loss: 0.1635 - accuracy: 0.9531 - val_loss: 0.0760 - val_accuracy: 0.9778
Epoch 2/10
748/748 [==============================] - 22s 30ms/step - loss: 0.0467 - accuracy: 0.9868 - val_loss: 0.0433 - val_accuracy: 0.9871
Epoch 3/10
748/748 [==============================] - 23s 31ms/step - loss: 0.0321 - accuracy: 0.9898 - val_loss: 0.0389 - val_accuracy: 0.9891
Epoch 4/10
748/748 [==============================] - 23s 30ms/step - loss: 0.0229 - accuracy: 0.9929 - val_loss: 0.0430 - val_accuracy: 0.9881
Epoch 5/10
527/748 [====================>.........] - ETA: 6s - loss: 0.0142 - accuracy: 0.9960

Certainly, here's a revised version with descriptions:

- The output encapsulates the neural network's performance metrics averaged across 10 training epochs.
  - The average training loss was calculated at 0.0397, indicating the model's mean error during training.
  - Training accuracy averaged at 99.3%, reflecting the percentage of correctly classified samples in the training dataset.
- Similarly, the average validation loss was observed to be 0.0432, denoting the model's average error on unseen validation data.
  - The validation accuracy averaged at 99.06%, illustrating the model's effectiveness in correctly classifying samples in the validation dataset.
  
These metrics collectively offer valuable insights into the model's learning behavior and its generalization capabilities to new, unseen data.

# Model Evaluation

In [ ]:
# Evaluate the model
test_loss, test_acc = model.evaluate(X_test, y_test)
print(f"Test Accuracy: {test_acc}")


Certainly, here's a concise paraphrase in two points:

The test accuracy, standing at around 99.130%, denotes the proportion of correctly identified typefaces within the test dataset. This metric highlights the model's effectiveness in classification tasks based on the provided data.

Conversely, the test loss, registered at 0.0425, signifies the model's predictive accuracy by quantifying the disparity between its predictions and the true labels. Lower test loss values generally indicate better alignment between predicted and actual outcomes, emphasizing the model's precision in typeface classification.

These evaluation metrics collectively indicate the trained model's robust performance and its ability to generalize well to unseen data.

In [ ]:
# Plot training & validation accuracy values
plt.plot(history.history['accuracy'])
plt.plot(history.history['val_accuracy'])
plt.title('Model accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Test'], loc='upper left')
plt.show()

# Plot training & validation loss values
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('Model loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Test'], loc='upper left')
plt.show()

**Model Accuracy Plot:**
- **Immediate Progress:** The training dataset swiftly achieves higher accuracy from the beginning.
- **Consistent Enhancement:** Training accuracy steadily improves, albeit with a diminishing rate of improvement as epochs progress.
- **Test Accuracy Pattern:** Test accuracy experiences an initial sharp rise, followed by a slower increase and eventual stabilization.
- **Reduction in Discrepancy:** Initially, there's a significant gap between training and test accuracies, which gradually diminishes, indicating better generalization.

**Model Loss Plot:**
- **Rapid Descent:** Training loss rapidly decreases in the initial epoch, signifying fast learning.
- **Gradual Reduction:** Training loss continues to decrease, although at a slower pace, as the model fine-tunes its parameters.
- **Test Loss Transition:** Test loss decreases notably but reaches a plateau earlier compared to training loss.
- **Convergence Stage:** Towards the end of training, both training and test losses stabilize, indicating the model has reached a stable state.

In [ ]:
# Make predictions
predictions = model.predict(X_test)

Upon execution of this code, the variable 'predictions' will hold an array of predicted probabilities for each test image, spanning all distinct typefaces. Subsequent steps would entail assessing and analyzing these predictions to gauge the model's performance.

In [ ]:
# Example: Print the predicted label for the first test image
predicted_label = np.argmax(predictions[0])
print(f"Predicted Label: {predicted_label}")

- Since the typefaces are represented by numerical labels, the index corresponds to the predicted typeface for the given image.
- In this specific example, the output indicates that the model predicts the label 3 for the first test image.

In [ ]:
# Visualize the first test image
plt.figure()
plt.imshow(X_test[0].reshape(28, 28), cmap='gray')
plt.title(f"True Label: {y_test[0]}, Predicted Label: {predicted_label}")
plt.show()

The above result shows that our predicted label was correct.

# Conclusion
- The CNN model achieved an impressive 99.23% test accuracy, indicating its suitability for tasks involving structured data like images and time-series data.
- Its architecture comprised convolutional layers for identifying local patterns, pooling layers for reducing data dimensionality while preserving crucial features, and fully connected layers for classification purposes.
- Nonlinear activation functions such as ReLU, sigmoid, and tanh were integrated to enable the model to understand complex data representations, enhancing its capability to detect intricate patterns within the input.


#References

Many techniques used in this notebook have been adopted from the following github repositories:

Author name - Prof Nik Bear Brown Link - https://github.com/nikbearbrown/

TMNIST (Typeface MNIST) - https://www.kaggle.com/nimishmagre/tmnist-typeface-mnistLinks to an external site.



# Licensing
MIT License

Copyright (c) 2024 AnuragGher20

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.